### PR TITLE
feat(Role): role tags

### PR DESCRIPTION
**Pull request by almostSouji**
**Please describe the changes this PR makes and why it should be merged:**

role data now includes tags, indicating if the role is the boost role (premium subscription in discord lingo), belongs to an integration or bot. This feature request aims to implement this new feature into discord.js while also introducing some useful getters

**Changes breakdown:**

- Role#tags introduce the new API data into a #tags property on Role
- ~~Role#isMyRole~~ *
- ~~Role#isPremiumSubscriberRole~~ *
- GuildMemberRoleManager#premiumSubscriberRole
- GuildMemberRoleManager#botRole
- RoleManager#premiumSubscriberRole
- RoleManager#botRoleFor *
- ~~RoleManager#myRole~~ *
- (+) Integration#Roles (external PR) (see consideration 9.)
- typings

\* _See pruning considerations below_

~~_The feature part dealing with integrations is only minimal and does not need to introduce new access or getters due to consideration 8._~~ see consideration 9.
 
<details>
<summary><b>Considerations</b></summary>

1. **Naming**: we use #premiumSince #premiumSubscriberCount for other boosting features. The API data key is `premium_subscriber`, so the camel-case version would be #premiumSubscriber, which however sounds more like a GuildMember than a Role instance, so the suffix -Role is added
2. **Naming**: #myRole and #isMyRole after Guild#me
3. **Implementation**: GuildMemberRolemanager#myRole could be argued as useful to find the bots role from Guild#me
4. **Implementation**: ~~Reflect~~ the `in` operator along with this rather cumbersome check and assignment are used to only create the keys on the #tags object that are actually represented in the role data payload, otherwise we'd end up with some nasty undefined or null values.
5. **API**: the value for the `premium_subscriber` key is null as per API spec, transformed to `true` here for convenience
6. **Implementation:** I was thinking about GuildMember#isPremiumSubscriber, but since it is tightly knit to the role, figured it should be moved to the role manager instead. I then figured that additional utility could come from returning the member instead of the boolean setup of `is*` or `has*`. Note: This is not equivalent to GuildMember#guild#roles#premiumSubscriberRole as it can be used to check for said role on this specific member!
7. **Naming**: RoleManager#botRoleFor akin to #permissionsFor and #botRole 
8. ~~**Implementation**: Integration identifiers are more of a nieche part here and do not need new access as [Integration#role](https://discord.js.org/#/docs/main/stable/class/Integration?scrollTo=role)already exists~~ (see 9.)
9. **Implementation**: Integrations can have multiple roles (level roles for twitch, for example) thusly Integration#roles has been introduced.

</details>

<details>
<summary><b>API pruning considerations</b></summary>

This API has become very convoluted, i don't think this is necessary or exactly beneficial and will trim it back with the following considerations:

- Role#tags ✔ fundamental
- Role#isMyRole ❌ was only used internally, feels clunky
- Role#isPremiumSubscriberRole ❌ was only used internally, feels clunky
- GuildMemberRoleManager#premiumSubscriberRole ✔ permission check
- GuildMemberRoleManager#botRole ✔ easy access for a certain bots role, makes #isMyRole largely irrelevant
- RoleManager#premiumSubscriberRole ✔ easy access for a guilds premium subscription role
- RoleManager#myRole ❌
- RoleManager#botRoleFor(UserResolvable) ✔ can act as more general replacement for #myRole
- Integration#roles ✔ Integration#role just covers one of the possibly many integration roles

</details>

<details>
<summary>Past caveats and roadblocks</summary>

- ~~🌊 **upstream**: requires documentation https://github.com/discord/discord-api-docs/pull/2230~~
- ~~🚧 **caveat:** Proper compatibility with partials will require #4636 to be merged, see: https://github.com/discordjs/discord.js/pull/4628#discussion_r454067795~~ #4636 has been reverted, accordingly https://github.com/discordjs/discord.js/pull/4628/commits/066eaa0d6104eb00ee880ed8f42c07400e5069a0 removes the bot check to be compatible with partials
- ~~✏ **draft**: the feature is available but undocumented, upstream documentation might yet change~~
- ~~🚧 **caveat**: As of today (2020/07/12) Discord does not send the `tags` field with the initial role data. 
The feature works as intended when used after `<Guild>.roles.fetch()` is called to fetch all roles. According to the discord developer server the feature is ["partially deployed"](https://paste.nomsy.net/edefeqimoh.md)~~
- ~~🚧 **update**: https://github.com/discord/discord-api-docs/issues/1537#issuecomment-733255329
Guild deploy is being worked on, meaning guilds will include this data on the initial payloads once done. Also meaning convenience getters as proposed here are probably the way to go and we do not need to individually fetch roles behind the scenes in order to make it work - or awkward info boxes to tell people to do so.~~

</details>

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/managers/RoleManager.js
Comment (NotSugden) [inline src/managers/RoleManager.js:146]:
```suggestion
   * @type {?Role}
```

File: typings/index.d.ts
Comment (NotSugden) [inline typings/index.d.ts:1253]:
```suggestion
    public tags: RoleTagData | null;
```

File: src/structures/Role.js
Comment (NotSugden) [inline src/structures/Role.js:114]: wouldn't this return null if there were no tags

File: src/structures/Role.js
Comment (NotSugden) [inline src/structures/Role.js:122]: same as above

File: src/managers/RoleManager.js
Comment (NotSugden) [inline src/managers/RoleManager.js:142]: should this return null if the role isn't found? or should it return undefined still

File: src/managers/GuildMemberRoleManager.js
Comment (NotSugden) [inline src/managers/GuildMemberRoleManager.js:83]: same as above

File: src/managers/RoleManager.js
Comment (almostSouji) [reply src/managers/RoleManager.js:142]: in my opinion getters should propagate the find value in that case (which returns undefined in v12). Which would be consistent with other similar API behavior (<https://discord.js.org/#/docs/main/stable/class/User?scrollTo=avatarURL>, to name one)
I'll let our core devs decide on that one

File: src/managers/RoleManager.js
Comment (NotSugden) [inline src/managers/RoleManager.js:150]:
```suggestion
    return this.cache.find(role => role.isMyRole);
```
typo missed in first review

File: src/structures/Role.js
Comment (NotSugden) [inline src/structures/Role.js:114]: this would return undefined on a role that has tags, but isn't a premium subscriber role

**Comment by Androz2091** [comment]
There is no way to get the role of a bot from a GuildMember instance, except for the client?
Something like:
```js
const infoEmbed = new MessageEmbed();
if(member.user.bot){
    infoEmbed.addField('Role created for this bot', member.roles.itsRole.toString());
}
// this could simplify userinfo commands
```
For the moment, this can be found using:
```js
const infoEmbed = new MessageEmbed();
if(message.author.bot){
    infoEmbed.addField('Role created for this bot', member.roles.cache.find((r) => r.tags.botID === member.id).toString());
}
```
Maybe I'm wrong but as a property exists to find the role of the client, it seems logical to me to have a getter to easily get the role of the other bots

**Comment by Papaia** [comment]
@Androz2091 
```js
const botRole = roles.find(r => r.tags?.botID === botID);
```
Another getter, especially if named `itsRole`, is redundant.
It would essentially mean that every GuildMember will have this getter, bot or not bot, and `itsRole` is very confusing.

**But**
It could be nice if integrations had a role getter for them.

**Comment by almostSouji** [comment]
- I can see `<RoleManager>.botRole(<UserResolvable>)` to make that slightly easier to access
- I personally prefer Integration#role over `<RoleManager>.integrationRole` , though the latter might also have its place for consistencies sake.

File: typings/index.d.ts
Comment (NotSugden) [inline typings/index.d.ts:1963]:
```suggestion
    public readonly myRole: Role | null;
    public readonly premiumSubscriberRole: Role | null;
```

File: src/structures/Role.js
Comment (Vaporox) [inline src/structures/Role.js:91]:
```suggestion
     * @property {?boolean} premiumSubcriberRole Whether this is the guild's premium subscription role
```
Small nitpick

File: src/structures/Role.js
Comment (Vaporox) [inline src/structures/Role.js:108]:
```suggestion
   * Whether this role is the guild's premium subscription role
```
Same as above

File: src/structures/Role.js
Comment (Vaporox) [inline src/structures/Role.js:117]:
```suggestion
   * Whether this role is the role associated with the logged in client user
```
Should be made consistent with [this](https://github.com/discordjs/discord.js/pull/4628/files#diff-1e631e07c0e6467138781b3eed92f077R145)

**Comment by almostSouji** [comment]
This API has become very convoluted, i don't this this is necessary or exactly beneficial and will trim it back with the following considerations:

- Role#tags ✔ fundamental
- Role#isMyRole ❌ was only used internally, feels clunky
- Role#isPremiumSubscriberRole ❌ was only used internally, feels clunky
- GuildMemberRoleManager#premiumSubscriberRole ✔ permission check
- GuildMemberRoleManager#botRole ✔ easy access for a certain bots role, makes #isMyRole largely irrelevant
- RoleManager#premiumSubscriberRole ✔ easy access for a guilds premium subscription role
- RoleManager#myRole ❌
- RoleManager#botRoleFor(UserResolvable) ✔ can act as more general replacement for #myRole

File: src/managers/GuildMemberRoleManager.js
Comment (NotSugden) [inline src/managers/GuildMemberRoleManager.js:91]:
```suggestion
    if (!this.client.user.bot) return null;
```
incase for whatever reason the bots GuildMember object is uncached

Comment (almostSouji) [reply src/managers/GuildMemberRoleManager.js:91]: No, the purpose of this is to target the member the GuildMemberRoleManager belongs to and is [not nullable](https://discord.js.org/#/docs/main/stable/class/GuildMemberRoleManager?scrollTo=member), this would check if the client user is not a bot, which is obsolete and not the point of the check.

Comment (NotSugden) [reply src/managers/GuildMemberRoleManager.js:91]: oh right, misread that, mb

File: src/managers/GuildMemberRoleManager.js
Comment (Vaporox) [inline src/managers/GuildMemberRoleManager.js:92]:
```suggestion
    return this.cache.find(role => role.tags && role.tags.botID === this.member.user.id) || null;
```

File: typings/index.d.ts
Comment (Vaporox) [inline typings/index.d.ts:2972]:
```suggestion
    premiumSubscriberRole?: true;
```
Would make more sense because it's either `true` or not there at all (also fixes typo)

**Comment by almostSouji** [comment]
@Androz2091 thanks for the suggestion, this is now in as `member.roles.botRole`. (the members bot role, getter)
If you want the role for other bots it's also available through `guild.roles.botRoleFor(botUser)`

I think this is now both usable as well as verbosely named.

I will also add another consideration note to the bunch above regarding #botRoleFor not being called on the #botRole getter

**Comment by almostSouji** [comment]
🚧 As of today Discord does not send the `tags` field with the initial role data. 
The feature works as intended when used after `<Guild>.roles.fetch()` is called to fetch all roles.

According to the discord developer server the feature is ["partially deployed"](https://paste.nomsy.net/edefeqimoh.md)

File: src/managers/RoleManager.js
Comment (NotSugden) [inline src/managers/RoleManager.js:134]: might this return null 'falsely' for a partial user?

Comment (almostSouji) [reply src/managers/RoleManager.js]: removed bot check in https://github.com/discordjs/discord.js/pull/4628/commits/066eaa0d6104eb00ee880ed8f42c07400e5069a0

File: src/managers/RoleManager.js
Comment (Vaporox) [inline src/managers/RoleManager.js:136]:
```suggestion
  botRoleFor(user) {
    user = this.client.users.resolveID(user);
    return this.cache.find(role => role.tags && role.tags.botID === user) || null;
  }
```
This way there's not even a cached user object needed, since we only need the ID to find the bot role.

Comment (almostSouji) [reply src/managers/RoleManager.js]: implemented variation in b1fe25c to still keep the fast return if no input is provided (the resolveID call seems to be neglectable) and increase verbosity by not naming the id "user"

**Comment by almostSouji** [comment]
🚧 **update**: https://github.com/discord/discord-api-docs/issues/1537#issuecomment-733255329
Guild deploy is being worked on, meaning guilds will include this data on the initial payloads once done. Also meaning convenience getters as proposed here are probably the way to go and we do not need to individually fetch roles behind the scenes in order to make it work - or awkward info boxes to tell people to do so.

**Comment by Alex-R-31** [comment]
https://github.com/discord/discord-api-docs/issues/1537#issuecomment-733942178 reply rollout is complete

**Comment by almostSouji** [comment]
If there is any feedback from contribs/maintainers regarding the proposed API or implementation do let me know here or on discord - as you prefer and expect interaction to be necessary.

**Comment by almostSouji** [comment]
documentation is imminent: https://github.com/discord/discord-api-docs/pull/2230

File: src/structures/Role.js
Comment (NotSugden) [inline src/structures/Role.js:91]:
```suggestion
     * @property {true} [premiumSubscriberRole] Whether this is the guild's premium subscription role
```
i think this is technically correct here, as the `?` indicates it can be null, at the least it should be marked optional with `[]` though

same for the 2 lines above

Comment (almostSouji) [reply src/structures/Role.js]: according to vlad this needs both ? as well as []

Comment (Vaporox) [reply src/structures/Role.js]: Why would it need both? It's not nullable, just optional

Comment (vladfrangu) [reply src/structures/Role.js]: Oh, true, welcome to 
image

Comment (almostSouji) [reply src/structures/Role.js]: same, resolved

**Review by kyranet — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Pull request merged**
